### ISSUE TITLE
Prefer default parameters

### ISSUE THREAD

**Issue by jamestalmage** [sindresorhus/eslint-plugin-unicorn#29]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2460%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/29)

Applies only to `es6` mode:

Bad:

``` js
function (foo) {
  foo = foo || 'bar';
  // ...
}
```

Good:

``` js
function (foo = 'bar') {
  // ...
}
```

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)

### Submitted pull Requests
- [#632 Add `prefer-default-parameters` rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/632)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by sindresorhus** [comment]
Too bad I can't use default parameters in reusable modules in the near future...

**Comment by jfmengels** [comment]
The bad and the good examples are pretty different actually, as default parameters only apply when `foo` is `undefined`, and the `||` operator when `foo` is falsy.
I agree that default parameters should be used instead of `||` when suitable, but `||` does have its use cases.

**Comment by IssueHuntBot** [comment]
@issuehunt has funded $60.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/29) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/29).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by MrHen** [comment]
It would be nice if this rule had an "always" or "never" configuration option.

**Comment by sindresorhus** [comment]
@MrHen Why would you want `never`? I'm genuinely curious.

**Comment by MrHen** [comment]
I've stopped using default parameters entirely because of `null` causing issues with destructuring.

```
function foo({ bar = 123 } = {}) {
    console.log(bar);
}

foo(null); // TypeError: Cannot destructure property `bar` of 'undefined' or 'null'.
```

----

```
function foo(foo) {
    const { bar = 123 } = (foo || {});
    console.log(bar);
}

foo(null); // 123
```

It's also super confusing using defaults with destructuring because there are different ways to do it with subtle differences:

```
function foo2({ bar } = { bar: 123 }) {
    console.log(bar);
}

foo2(); // 123
foo2({}); // undefined

function foo3({ bar = 123 } = {}) {
    console.log(bar);
}

foo3(); // 123
foo3({}); // 123

function foo4({ bar = 123 } = { bar: 456 }) {
    console.log(bar);
}

foo4(); // 456
foo4({}); // 123
```

I guess what I _really_ care about is not using destructuring in function parameter defaults. Maybe that should just be a different rule.

**Comment by sindresorhus** [comment]
> I've stopped using default parameters entirely because of null causing issues with destructuring.

There's a solution to that: [I have completely stopped using `null`.](https://github.com/sindresorhus/meta/issues/7)

> It's also super confusing using defaults with destructuring because there are different ways to do it with subtle differences:

I would argue the only correct way to do it is this variant `function foo2({ bar } = { bar: 123 }) {`.

> I guess what I really care about is not using destructuring in function parameter defaults. Maybe that should just be a different rule.

Or maybe enforce only using `function foo2({ bar } = { bar: 123 }) {`?

**Comment by MrHen** [comment]
> There's a solution to that: I have completely stopped using null.

Sure, but from a defensive programming standpoint, I'd still rather not blow up on `null`.

> I would argue the only correct way to do it is this variant function foo2({ bar } = { bar: 123 }) {.

If the intent is to provide a default value for `bar`, then it would be better to use `{ bar = 123 } = {}`. Think of a standard options block:

```
function foo1({optionA = true, optionB = true} = {}) {
}
foo1({ optionB: false }); // preserves optionA = true

function foo2({optionA, optionB} = {optionA: true, optionB: true}) {
}
foo2({ optionB: false }); // optionA is now undefined and falsey
```

> Or maybe enforce only using function foo2({ bar } = { bar: 123 }) {?

I guess for the scope of this issue / rule, which of these statements are true / approved?

- [x] Disallow `function (foo) { foo = foo || 'bar'; }`
- [ ] Disallow `function (foo) { const bar = foo || 'bar'; }`
- [x] Allow `function (foo = 'bar') { }`
- [ ] Allow `function (foo = { bar: 123 }) { }`
- [ ] Allow `function ({ bar } = { bar: 123 }) { }`
- [ ] Allow `function ({ bar = 123 } = { bar }) { }`
- [ ] Allow `function ({ bar = 123 } = { }) { }`
- [ ] Allow `function (foo) { const { bar } = (foo || {}); }`

**Comment by sindresorhus** [comment]
Ugh, I copied the wrong code... I meant to copy `function foo3({ bar = 123 } = {}) {`. That is what I think is the correct solution.

**Comment by MrHen** [comment]
I just realized that https://github.com/sindresorhus/eslint-plugin-unicorn/issues/208 is also open and has a similar discussion.

**Comment by MrHen** [comment]
Alright, between the two tickets covering this area, here is my recommendation:

## `no-default-parameter` should

- Allow `function (foo) { foo = foo || 'bar'; }`
- Allow `function (foo) { const bar = foo || 'bar'; }`
- Allow `function (foo) { const { bar } = (foo || {}); }`
- Disallow `function (foo = { bar: 123 }) { }`
with fix to `function (foo: { bar = 123 } = {}) { }`
- Disallow `function ({ bar } = { bar: 123 }) { }`
with fix to `function ({ bar = 123 } = {}) { }`
- Disallow `function ({ bar = 123 } = { bar }) { }`
with fix to `function ({ bar = 123 } = {}) { }`
- Disallow `function (foo = fooDefault) { }`
without fix
- Allow `function (foo = {}) { }`
- Allow `function (foo = 'bar') { }`
- Allow `function ({ bar = 123 } = { }) { }`

## `prefer-default-parameters` should

- Disallow `function (foo) { foo = foo || 'bar'; }`
with fix to `function (foo = 'bar') { }`
- Disallow `function (foo) { const bar = foo || 'bar'; }`
with fix to `function (bar = 'bar') { }`
- Disallow `function (foo) { const { bar } = (foo || {}); }`
with fix to `function ({ bar } = {}) { }`
- Allow `function (foo = { bar: 123 }) { }`
- Allow `function ({ bar } = { bar: 123 }) { }`
- Allow `function ({ bar = 123 } = { bar }) { }`
- Allow `function (foo = fooDefault) { }`
- Allow `function (foo = {}) { }`
- Allow `function (foo = 'bar') { }`
- Allow `function ({ bar = 123 } = { }) { }`

----

That being said, I prefer the exact opposite of both of these rules so maybe I'm not the best person to make a call on this.

**Comment by JLHwung** [comment]
The `prefer-default-parameters` fix would be tricky in some cases since the default parameter is evaluated in the implicit intermediate scope. Consider the following example
```js
let a = "outside";
function f(b) {
  let a = "inside";
  b = b || a;
  return b;
}
f() // inside
```

The fixed code would change the semantics and print "outside" when evaluating `f()` since `() => a` is evaluated in parameter scope.

I suggest we limit the fixer capacity that we only fix those with `literal` as initializer. For those expression initializer, we should report error `prefer-default-parameters` and do not try to fix this.

**Comment by sindresorhus** [comment]
> Sure, but from a defensive programming standpoint, I'd still rather not blow up on null.

I would argue blowing up on `null` is the most defensive thing you can do. I think you meant to say "graceful", which is ok too, I just personally dread `null`.

---

> no-default-parameter

I don't really want this. Default parameters are not going away. I don't think it makes sense to fight it.

I would prefer the rule to be named `prefer-default-parameter` and only enforce default parameters.

---

> prefer-default-parameters should

:+1: I agree with the Allow/Disallow's here.

**Comment by sindresorhus** [comment]
> I suggest we limit the rule and fixer capacity that we only fix those with literal as initializer. For those expression initializer, we should not report prefer-default-parameters error and of course, not try to fix that.

I agree with not fixing such cases, but why can we not report? The user could still rework it to work as a default parameter.

**Comment by MrHen** [comment]
> I don't really want this. Default parameters are not going away. I don't think it makes sense to fight it.

Should we close the other issue then? Personally, I would use `no-default-parameter` over `prefer-default-parameters` but we obviously can't have both rules default to `error`. In terms of priority, I think I agree with you that `prefer-default-parameters` is a better fit for `eslint-plugin-unicorn`.

**Comment by fregante** [comment]
@MrHen this doesn't seem to be valid ES:

```js
function (foo: { bar = 123 } = {}) { }
```
